deepspeed/runtime/engine.py

```diff
@@ -1742,6 +1745,8 @@ def allreduce_bucket(self, bucket, dp_group):
 
         if self.allreduce_always_fp32():
             tensor_to_allreduce = tensor.float()
+        elif self.allreduce_always_fp16():
+            tensor_to_allreduce = tensor.half()
```

Comment:
Does this mean that it is possible to configure fp32 training with fp16 reduction? I can't think of scenarios where that might be useful? Can you please share more details on the motivation for this PR?

Reply:
I have a very large model (5-10 GB of weights) on the backward step when we try to synchronize gradients we need to transfer whole this weights between GPUs. If we have no NVLINK and we want to use fp32 weights to train we will be limited by PCIe bandwidth (5-10 GB/s for PCIe3 and 10-20 GB/s for PCIe4). Transfering gradients in fp16 and after that applying all arithmetic operations using fp32 we do not lose a lot of quality.

Reply:
That makes sense. I was not aware of such scenarios, thanks for expanding my knowledge. The only concern I have is that we now have two conflicting configuration flags for specifying reduction data type. It would be preferable to consolidate the reduction data type into a single configuration flag:
```json
{
  "reduction_type": "fp16 | fp32 | bfloat16 | int8, ..."
}
```
What do you think?

Reply:
Sounds logical. I will rewrite the code to use a single flag.